File: src/main/java/org/pyjjs/scheduler/core/data/ObservableDataSource.java
package org.pyjjs.scheduler.core.data;

import com.google.common.collect.ArrayListMultimap;
import com.google.common.collect.Multimap;
import org.pyjjs.scheduler.core.model.IdentifiableObject;

import java.util.*;
import java.util.function.Consumer;

public abstract class ObservableDataSource implements Collection<IdentifiableObject> {

    private Multimap<Class<? extends IdentifiableObject>, DataSourceListener> dataSourceListeners;

    public ObservableDataSource() {
        dataSourceListeners = ArrayListMultimap.create();
    }

    public void addDataSourceListener(DataSourceListener dataSourceListener) {
        dataSourceListeners.put(IdentifiableObject.class, dataSourceListener);
    }

    public void removeDataSourceListener(DataSourceListener dataSourceListener) {
        dataSourceListeners.remove(IdentifiableObject.class, dataSourceListener);
    }

    public void addDataSourceListener(Class<? extends IdentifiableObject> interestType, DataSourceListener dataSourceListener) {
        dataSourceListeners.put(interestType, dataSourceListener);
    }

    public void removeDataSourceListener(Class<? extends IdentifiableObject> interestType, DataSourceListener dataSourceListener) {
        dataSourceListeners.remove(interestType, dataSourceListener);
    }

    public void addDataSourceListener(List<Class<? extends IdentifiableObject>> interestTypes, DataSourceListener dataSourceListener) {
        for (Class<? extends IdentifiableObject> interestType : interestTypes) {
            dataSourceListeners.put(interestType, dataSourceListener);
        }
    }

    public void removeDataSourceListener(List<Class<? extends IdentifiableObject>> interestTypes, DataSourceListener dataSourceListener) {
        for (Class<? extends IdentifiableObject> interestType : interestTypes) {
            dataSourceListeners.remove(interestType, dataSourceListener);
        }
    }

    public void notifyAboutCreate(IdentifiableObject entity) {
        notify(entity, (listener) -> listener.onCreate(entity));
    }

    public void notifyAboutUpdate(IdentifiableObject entity) {
        notify(entity, (listener) -> listener.onUpdate(entity));
    }

    public void notifyAboutRemove(IdentifiableObject entity) {
        notify(entity, (listener) -> listener.onRemove(entity));
    }

    private void notify(IdentifiableObject entity, Consumer<DataSourceListener> notification) {
        dataSourceListeners.entries()
                .stream()
                .forEach(listenerEntry -> {
                    if (listenerEntry.getKey().isAssignableFrom(entity.getClass())) {
                        notification.accept(listenerEntry.getValue());
                    }
                });
    }

    public interface DataSourceListener {

        void onCreate(IdentifiableObject entity);

        void onUpdate(IdentifiableObject entity);

        void onRemove(IdentifiableObject entity);

    }
}

File: src/main/kotlin/org/pyjjs/scheduler/core/data/ObservableDataSource.kt
package org.pyjjs.scheduler.core.data

import com.google.common.collect.ArrayListMultimap
import com.google.common.collect.Multimap
import org.pyjjs.scheduler.core.model.IdentifiableObject

abstract class ObservableDataSource : MutableCollection<IdentifiableObject> {

    private val dataSourceListeners: Multimap<Class<out IdentifiableObject>, DataSourceListener>

    init {
        dataSourceListeners = ArrayListMultimap.create<Class<out IdentifiableObject>, DataSourceListener>()
    }

    fun addDataSourceListener(dataSourceListener: DataSourceListener) {
        dataSourceListeners.put(IdentifiableObject::class.java, dataSourceListener)
    }

    fun removeDataSourceListener(dataSourceListener: DataSourceListener) {
        dataSourceListeners.remove(IdentifiableObject::class.java, dataSourceListener)
    }

    fun addDataSourceListener(interestType: Class<out IdentifiableObject>, dataSourceListener: DataSourceListener) {
        dataSourceListeners.put(interestType, dataSourceListener)
    }

    fun removeDataSourceListener(interestType: Class<out IdentifiableObject>, dataSourceListener: DataSourceListener) {
        dataSourceListeners.remove(interestType, dataSourceListener)
    }

    fun addDataSourceListener(interestTypes: List<Class<out IdentifiableObject>>, dataSourceListener: DataSourceListener) {
        for (interestType in interestTypes) {
            dataSourceListeners.put(interestType, dataSourceListener)
        }
    }

    fun removeDataSourceListener(interestTypes: List<Class<out IdentifiableObject>>, dataSourceListener: DataSourceListener) {
        for (interestType in interestTypes) {
            dataSourceListeners.remove(interestType, dataSourceListener)
        }
    }

    fun notifyAboutCreate(entity: IdentifiableObject) {
        notify(entity, { listener -> listener.onCreate(entity) })
    }

    fun notifyAboutUpdate(entity: IdentifiableObject) {
        notify(entity, { listener -> listener.onUpdate(entity) })
    }

    fun notifyAboutRemove(entity: IdentifiableObject) {
        notify(entity, { listener -> listener.onRemove(entity) })
    }

    private fun notify(entity: IdentifiableObject, notification: (DataSourceListener) -> Unit) {
        dataSourceListeners.entries().forEach { listenerEntry ->
            if (listenerEntry.key.isAssignableFrom(entity.javaClass)) {
                notification.invoke(listenerEntry.value)
            }
        }
    }

    interface DataSourceListener {

        fun onCreate(entity: IdentifiableObject)

        fun onUpdate(entity: IdentifiableObject)

        fun onRemove(entity: IdentifiableObject)

    }
}
